<a href="https://colab.research.google.com/github/PikuFuka/CSEL-302-comp/blob/main/2B_SOTOMAYOR_EXER5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Importing Libraries**

In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.preprocessing import StandardScaler

# **1. Data Preprocessing:**


In [ ]:
df = pd.read_csv("BankCustomerData.csv")
df

age           job  marital  education default  balance housing loan  \
0       58    management  married   tertiary      no     2143     yes   no   
1       44    technician   single  secondary      no       29     yes   no   
2       33  entrepreneur  married  secondary      no        2     yes  yes   
3       47   blue-collar  married    unknown      no     1506     yes   no   
4       33       unknown   single    unknown      no        1      no   no   
...    ...           ...      ...        ...     ...      ...     ...  ...   
42634   21       student   single  secondary      no     2488      no   no   
42635   87       retired  married    primary      no     2190      no   no   
42636   34   blue-collar  married    primary      no     6718      no   no   
42637   22       student   single  secondary      no      254      no   no   
42638   32    management   single   tertiary      no     1962      no   no   

         contact  day month  duration  campaign  pdays  previous poutcome  \
0        unknown    5   may       261         1     -1         0  unknown   
1        unknown    5   may       151         1     -1         0  unknown   
2        unknown    5   may        76         1     -1         0  unknown   
3        unknown    5   may        92         1     -1         0  unknown   
4        unknown    5   may       198         1     -1         0  unknown   
...          ...  ...   ...       ...       ...    ...       ...      ...   
42634  telephone   12   jan       661         2     92         1  success   
42635  telephone   12   jan       512         2     -1         0  unknown   
42636   cellular   13   jan       278         4     97         1    other   
42637   cellular   13   jan       143         2     -1         0  unknown   
42638   cellular   13   jan       130         1     -1         0  unknown   

      term_deposit  
0               no  
1               no  
2               no  
3               no  
4               no  
...            ...  
42634          yes  
42635          yes  
42636           no  
42637          yes  
42638           no  

[42639 rows x 17 columns]

# **Dummies and Convertion**

In [ ]:
df['term_deposit'] = df['term_deposit'].map({'yes': 1, 'no': 0})
df['default'] = df['default'].map({'yes': 1, 'no': 0})
df['housing'] = df['housing'].map({'yes': 1, 'no': 0})
df['loan'] = df['loan'].map({'yes': 1, 'no': 0})

# **Dummies**

In [ ]:
df_dumies = pd.get_dummies(df, drop_first=True)
df_dumies['subscribed'] =  (df['term_deposit']>0).astype(int)
df_dumies

age  default  balance  housing  loan  day  duration  campaign  pdays  \
0       58        0     2143        1     0    5       261         1     -1   
1       44        0       29        1     0    5       151         1     -1   
2       33        0        2        1     1    5        76         1     -1   
3       47        0     1506        1     0    5        92         1     -1   
4       33        0        1        0     0    5       198         1     -1   
...    ...      ...      ...      ...   ...  ...       ...       ...    ...   
42634   21        0     2488        0     0   12       661         2     92   
42635   87        0     2190        0     0   12       512         2     -1   
42636   34        0     6718        0     0   13       278         4     97   
42637   22        0      254        0     0   13       143         2     -1   
42638   32        0     1962        0     0   13       130         1     -1   

       previous  ...  month_jun  month_mar  month_may  month_nov  month_oct  \
0             0  ...          0          0          1          0          0   
1             0  ...          0          0          1          0          0   
2             0  ...          0          0          1          0          0   
3             0  ...          0          0          1          0          0   
4             0  ...          0          0          1          0          0   
...         ...  ...        ...        ...        ...        ...        ...   
42634         1  ...          0          0          0          0          0   
42635         0  ...          0          0          0          0          0   
42636         1  ...          0          0          0          0          0   
42637         0  ...          0          0          0          0          0   
42638         0  ...          0          0          0          0          0   

       month_sep  poutcome_other  poutcome_success  poutcome_unknown  \
0              0               0                 0                 1   
1              0               0                 0                 1   
2              0               0                 0                 1   
3              0               0                 0                 1   
4              0               0                 0                 1   
...          ...             ...               ...               ...   
42634          0               0                 1                 0   
42635          0               0                 0                 1   
42636          0               1                 0                 0   
42637          0               0                 0                 1   
42638          0               0                 0                 1   

       subscribed  
0               0  
1               0  
2               0  
3               0  
4               0  
...           ...  
42634           1  
42635           1  
42636           0  
42637           1  
42638           0  

[42639 rows x 44 columns]

# **2. Feature Selection:**

In [ ]:
X = df_dumies.drop(['default', 'balance', 'housing', 'loan', 'term_deposit', 'subscribed'], axis=1)
y = df_dumies['subscribed']

# **3. Data Splitting:**

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# **Standarize the features**

In [ ]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.fit_transform(X_test)

# **4. Model Training:**

In [ ]:
model = LogisticRegression()
model.fit(X_train_scaled, y_train)

LogisticRegression()

# **5. Model Evaluation:**

In [ ]:
y_pred = model.predict(X_test_scaled)

# **Evaluate the model**

In [ ]:
accuracy = accuracy_score(y_test, y_pred)
conf_matrix = confusion_matrix(y_test, y_pred)
class_report = classification_report(y_test, y_pred)

print(f"Accuracy: {accuracy}")
print("confusion matrix: ")
print(conf_matrix)
print("classification report: ")
print(class_report)

Accuracy: 0.9179174484052532
confusion matrix: 
[[7593  135]
 [ 565  235]]
classification report: 
              precision    recall  f1-score   support

           0       0.93      0.98      0.96      7728
           1       0.64      0.29      0.40       800

    accuracy                           0.92      8528
   macro avg       0.78      0.64      0.68      8528
weighted avg       0.90      0.92      0.90      8528



# **6. Conclusion:**

The model accuracy is 0.92.
Confusion Matrix:

True Positives (TP): 7593
False Positives (FP): 135
True Negatives (TN): 235
False Negatives (FN): 565

It's balanced, especially in catching instances of the smaller class. While it still makes some mistakes, overall, it's doing a good job.
